pkg/asset/imagebased/image/baseiso.go

```diff
@@ -0,0 +1,128 @@
+package image
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/coreos/stream-metadata-go/arch"
+	"github.com/coreos/stream-metadata-go/stream"
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/rhcos"
+	"github.com/openshift/installer/pkg/rhcos/cache"
+	"github.com/openshift/installer/pkg/types"
+)
+
+var (
+	defaultCoreOSStreamGetter = rhcos.FetchCoreOSBuild
+)
+
+// BaseIso generates the base ISO file for the image.
+type BaseIso struct {
+	File *asset.File
+
+	streamGetter CoreOSBuildFetcher
+}
+
+// CoreOSBuildFetcher will be to used to switch the source of the coreos metadata.
+type CoreOSBuildFetcher func(ctx context.Context) (*stream.Stream, error)
+
+var _ asset.WritableAsset = (*BaseIso)(nil)
+
+// Name returns the human-friendly name of the asset.
+func (i *BaseIso) Name() string {
+	return "BaseIso Image"
+}
+
+// Dependencies returns dependencies used by the asset.
+func (i *BaseIso) Dependencies() []asset.Asset {
+	return []asset.Asset{}
+}
+
+// Generate generates the base ISO.
+func (i *BaseIso) Generate(_ context.Context, dependencies asset.Parents) error {
```

Comment:
IIUC It looks like there is a good amount of code duplication to skip a portion of the original [BaseIso.retrieveBaseIso](https://github.com/openshift/installer/blob/54e81ef067d432cc41a3333765b2f28ad34bbbea/pkg/asset/agent/image/baseiso.go#L208-L233) and avoid the not required dependencies. Probably least the [downloadIso](https://github.com/openshift/installer/blob/54e81ef067d432cc41a3333765b2f28ad34bbbea/pkg/asset/agent/image/baseiso.go#L74-L93) method could be shared.

Reply:
That's right, I decided on duplicating code rather than trying to share as much as I could, because what this piece is simple enough and to not introduce changes in the agent pkg. But I agree that sharing at least the `downloadIso` method is quite simple and leaves us with one less method to have as a duplicate.

Would you export the `downloadIso` method in the agent pkg and import the latter here?

Reply:
Probably the `downloadIso` could be even refactored out into its own type (ie something like `IsoDownloader`) as being pretty stand-alone, and reused where required

Reply:
That sounds even better, I'll try to propose the respective change.